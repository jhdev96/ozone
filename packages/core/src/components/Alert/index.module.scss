@import '../../scss/root';
@import '../../scss/mixins/alert/alert.scss';


/* Main alert classes */
.progress {
  @include alert;

  &--light {
    color: var(--black-core);
    border: 1px solid var(--gray-1);
    background-color: #fff;
  }
  &--dark {
    color: var(--gray-1);
    border: 1px solid var(--black-5);
    background-color: var(--black-3) !important;
  }
}

.info {
  @include alert;

  color: var(--blue-6);
  background-color: var(--blue-2);
}

.error {
  @include alert;

  color: var(--red-4);
  background-color: var(--red-0);
}

.warning {
  @include alert;

  color: var(--yellow-4);
  background-color: var(--yellow-1);
}

.success {
  @include alert;

  color: var(--green-3);
  background-color: var(--green-1);

  &__icon {
    color: #215F0C;
  }
}

.mssg {
  @extend .alert__mssg;
}

.closeBtn {
  @extend .alert__close-btn;
}

.animate {
  @include floatAnimation;
}