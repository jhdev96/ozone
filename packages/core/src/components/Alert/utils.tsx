import React from 'react';
import LoadingDots from '../LoadingDots';
import { Info } from '@jhdev96/ozone-icons';
import { Warning } from '@jhdev96/ozone-icons';
import { Success } from '@jhdev96/ozone-icons';
import { Error } from '@jhdev96/ozone-icons';
import Colors from '@jhdev96/ozone-colors';


export type AlertVariation =  "progress"  | 
                              "info"      | 
                              "warning"   | 
                              "success"   | 
                              "error";

export const IconColor = {
  success: Colors.extended.GREEN2,
  error: Colors.extended.RED2,
  info: Colors.extended.BLUE5,
  warning: Colors.extended.YELLOW2,
  progress: Colors.grayScale.BLACK2
}

export const alertIcon = (
  Type: AlertVariation, 
  theme?: 'light' | 'dark'
): JSX.Element => {
  /**
   * Given an alertType and a color theme, 
   * return the appropriate icon for the alert.
   */
  theme ||= 'light';

  switch (Type) {
    case 'success': 
      return (
        <Success 
          color={IconColor.success} 
          className="successIcon"
        />
      );
    case 'warning': 
      return (
        <Warning 
          color={IconColor.warning} 
          className="warningIcon"
        />
      );
    case 'error': 
      return (
        <Error 
          color={IconColor.error} 
          className="errorIcon"
        />
      );
    case 'progress':
      return (
        <LoadingDots 
          theme={
            theme === 'light' 
              ? 'dark' 
              : 'light'
          } 
          position="left" 
          dotSize={6} 
          className="progressIcon"
        />
      );

    default: 
      return (
        <Info 
          color={IconColor.info} 
          className="infoIcon" 
        />
      );
  }
}