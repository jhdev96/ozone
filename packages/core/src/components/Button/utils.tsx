import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}