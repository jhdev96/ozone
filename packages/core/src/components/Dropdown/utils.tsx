import React from 'react';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export const menuOptions = (
  selectedOption: string, 
  options: (string | React.ReactNode)[], 
  onClick: (e: React.MouseEvent<HTMLDivElement>) => void
): React.ReactNode => {
  /**
    * Return an array of div elements
    * based on the options given by 
    * the client. Each div should 
    * have an onClick prop specifying
    * the function to be invoked when
    * the div or its child element is
    * clicked.
    */
  const optionDivs = options.map((
    option: string | React.ReactNode, 
    idx: React.Key
  ) => (
    <div 
      onClick={onClick} 
      className={styles["menu__option"]} 
      key={idx}
    >
      <span
        className={classNames(
          option === selectedOption && styles.selected
        )}
      >
        {option}
      </span>
    </div>
  ));

  return optionDivs;
};