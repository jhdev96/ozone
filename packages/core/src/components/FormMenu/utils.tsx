import React from 'react';
import styles from './index.module.scss';


export const createTransition = (
  slider: React.MutableRefObject<HTMLDivElement | undefined>, 
  delta: number
) => {
  /** Add a transition to the menu slider when moving to an option
   * that is more than 50 pixels away from the current option.
   * 
   * @param slider // a reference to the menu slider element
   * @param delta // the difference between the two options
   */
  if (Math.abs(delta) > 50) {
    slider.current!.classList.add(styles.transition);

    setTimeout(() => {
      slider.current!.classList.remove(styles.transition)
    }, 400);
  }
}