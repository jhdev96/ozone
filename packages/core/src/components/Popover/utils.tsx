import React from 'react';


const ARROW_SIZE = 10;
export type Position = "top" | "bottom" | "left" | "right";

export const positionPopper = (
  targetRef: React.MutableRefObject<HTMLElement | undefined>, 
  popperRef: React.MutableRefObject<HTMLElement | undefined>, 
  position?: Position
) =>  {
  /**
   * This function helps to position the popover
   * component relative to its target element.
   * @param targetRef // a ref pointing to the target element.
   * @param popperRef // a ref pointing to the popover.
   * @param position // where the popover should be place.
   */
  const target = targetRef.current;
  const popper = popperRef.current;

  switch (position) {
    case 'top':
      if (popper && target) {
        popper!.style.bottom = `${target!.offsetHeight + ARROW_SIZE}px`;
      }
      break;
    case 'left':
    case 'right':
      break;
    default:
      // bottom position
      if (popper && target) {
        popper!.style.top = `${target!.offsetHeight + ARROW_SIZE}px`;
      }
  }
}