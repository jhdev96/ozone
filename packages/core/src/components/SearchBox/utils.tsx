import React from 'react';
import styles from './index.module.scss';


export const toggleFocus = (e: React.FocusEvent<HTMLDivElement>) => {
  const classList = e.currentTarget.classList;
  
  if (classList.contains(styles.focus)) {
    classList.remove(styles.focus);
  } else {
      classList.add(styles.focus);
  }
}