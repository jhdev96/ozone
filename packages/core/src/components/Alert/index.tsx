import React, { useState } from 'react';
import { 
  IconColor, 
  alertIcon, 
  AlertVariation 
} from './utils';
import classNames from '../../utils/classNames';
import { ThemeType } from '../../utils/theme';
import { Close } from 'icons';
import styles from './index.module.scss';


export interface AlertProps {
  /** 
   * The type of alert being delivered. 
   */
  variation?: AlertVariation;
  /** 
   * The theme only applies for progress alerts. 
   */
  theme?: ThemeType;
  /** 
   * Determines if the user can close the alert. 
   */
  canBeClosed?: boolean; 
  /** 
   * Callback to be invoked when the alert is closed. 
   */
  onClose?: (e?: React.MouseEvent<HTMLDivElement>) => void;
  /**
   * If true, a floating animation will be applied
   * to the alert.
   */
  animate?: boolean;
  /** 
   * Selector used in testing environments. 
   */ 
  dataTestId?: string;
  /** 
   * The message rendered inside the alert. 
   */
  children?: string;
}

export default function Alert({
  variation     = "info", 
  theme         = "light",
  canBeClosed   = true,
  onClose       = () => {},
  animate       = false,
  dataTestId,
  children
}: AlertProps): JSX.Element | null
{
  const [visible, setVisible] = useState(true);
  const handleClose = (e: React.MouseEvent<HTMLDivElement>) => {
    setVisible(false);
    onClose(e);
  }

  if (visible) {
    return (
      <div 
        role="alert"
        className={classNames(
          styles[variation],
          styles[`${variation}--${theme}`],
          animate && styles.animate
        )}
        data-testid={dataTestId}
      >
        {alertIcon(variation, theme)}
        <div className={styles.mssg}>
          {children}
        </div>
        {canBeClosed && (
          <div 
            className={styles.closeBtn}
            onClick={handleClose}
          >
            <Close color={IconColor[variation]} />
          </div>
        )}
      </div>
    )
  }
  return null;
}

Alert.displayName = 'Alert';