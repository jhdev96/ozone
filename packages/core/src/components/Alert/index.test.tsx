import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Alert from '.';
import { AlertVariation } from './utils';


describe('<Alert />', () => {
  function renderAlerts() {
    const alertVariations: AlertVariation[] = [
      "progress",
      "info",
      "warning",
      "success",
      "error"
    ];

    for (let i = 0; i < alertVariations.length; i++) {
      render(
        <Alert 
          variation={alertVariations[i]} 
          dataTestId={alertVariations[i]}
        >
          Message
        </Alert>
      );
    }
  }

  it('matches snapshots', () => {
    renderAlerts();

    expect(screen.getByTestId('progress')).toMatchSnapshot();
    expect(screen.getByTestId('info')).toMatchSnapshot();
    expect(screen.getByTestId('warning')).toMatchSnapshot();
    expect(screen.getByTestId('success')).toMatchSnapshot();
    expect(screen.getByTestId('error')).toMatchSnapshot();
  });

  it('renders an Alert with a message', () => {
    const message = 'This is a warning';

    render(
      <Alert variation="warning" dataTestId="alert">
        {message}
      </Alert>
    );

    expect(screen.getByTestId('alert').textContent).toBe(message);
  });

  it('renders an alert with a close button', () => {
    const {container} = render(
      <Alert variation="info">Message</Alert>
    );

    expect(container.querySelector('.alertCloseBtn'))
      .toBeInTheDocument;
  });

  it('calls `onClose` when the alert is closed', () => {
    const onClose = jest.fn();
    const {container} = render(
      <Alert onClose={onClose}>Message</Alert>
    );
    const CloseButton = container.querySelector('.closeBtn');

    fireEvent.click(CloseButton!);
    expect(onClose).toBeCalled();
  });

  it('removes the alert from the DOM when `onClose` is clicked', () => {
    const onClose = jest.fn();
    const {container} = render(
      <Alert onClose={onClose}>Message</Alert>
    );
    const CloseButton = container.querySelector('.closeBtn');

    fireEvent.click(CloseButton!);
    expect(container).toBeEmptyDOMElement;
  });

  it('renders a progress alert with a dark theme', () => {
    render(
      <Alert variation="progress" theme="dark" dataTestId="progress">
        Message
      </Alert>
    );

    expect(screen.getByTestId('progress')).toHaveClass('progress--dark');
    expect(screen.getByTestId('progress').firstChild).toHaveClass('light');
  });

  it('renders an animated alert', () => {
    render(
      <Alert animate dataTestId="animated">Message</Alert>
    );

    expect(screen.getByTestId('animated')).toHaveClass('animate');
  });
});