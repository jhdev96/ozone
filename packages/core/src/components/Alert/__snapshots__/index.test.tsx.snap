// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> matches snapshots 1`] = `
<div
  class="progress progress--light"
  data-testid="progress"
  role="alert"
>
  <div
    class="progressIcon loader dark left"
  >
    <div
      class="top"
    >
      <div
        class="box"
        style="opacity: 1; min-height: 6px; min-width: 6px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 6px; min-width: 6px;"
      />
    </div>
    <div
      class="bottom"
    >
      <div
        class="box"
        style="opacity: 0; min-height: 6px; min-width: 6px;"
      />
      <div
        class="box"
        style="opacity: 0; min-height: 6px; min-width: 6px;"
      />
    </div>
  </div>
  <div
    class="mssg"
  >
    Message
  </div>
  <div
    class="closeBtn"
  >
    <svg
      height="10"
      id="Cancel"
      viewBox="0 0 15.021 15.021"
      width="7"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.374,14.5l-4.95-4.95L2.475,14.5A1.5,1.5,0,0,1,.354,12.374L5.3,7.425.354,2.475A1.5,1.5,0,0,1,2.475.354L7.425,5.3l4.95-4.95A1.5,1.5,0,0,1,14.5,2.475l-4.95,4.95,4.95,4.95A1.5,1.5,0,1,1,12.374,14.5Z"
        fill="#7F808A"
        id="Union_3"
        transform="translate(0.086 0.086)"
      />
    </svg>
  </div>
</div>
`;

exports[`<Alert /> matches snapshots 2`] = `
<div
  class="info info--light"
  data-testid="info"
  role="alert"
>
  <svg
    class="infoIcon"
    height="15"
    id="info-icon"
    viewBox="0 0 4 15"
    width="4"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="2"
      cy="2"
      fill="#236CB7"
      id="Ellipse_297"
      r="2"
    />
    <path
      d="M0,9.559H1A1.957,1.957,0,0,0,3,7.647V0H0Z"
      fill="#236CB7"
      id="Rectangle_1532"
      transform="translate(0.5 5.441)"
    />
  </svg>
  <div
    class="mssg"
  >
    Message
  </div>
  <div
    class="closeBtn"
  >
    <svg
      height="10"
      id="Cancel"
      viewBox="0 0 15.021 15.021"
      width="7"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.374,14.5l-4.95-4.95L2.475,14.5A1.5,1.5,0,0,1,.354,12.374L5.3,7.425.354,2.475A1.5,1.5,0,0,1,2.475.354L7.425,5.3l4.95-4.95A1.5,1.5,0,0,1,14.5,2.475l-4.95,4.95,4.95,4.95A1.5,1.5,0,1,1,12.374,14.5Z"
        fill="#236CB7"
        id="Union_3"
        transform="translate(0.086 0.086)"
      />
    </svg>
  </div>
</div>
`;

exports[`<Alert /> matches snapshots 3`] = `
<div
  class="warning warning--light"
  data-testid="warning"
  role="alert"
>
  <svg
    class="warningIcon"
    height="15"
    id="warning-icon"
    viewBox="0 0 4 15"
    width="4"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="2"
      cy="2"
      fill="#714601"
      id="Ellipse_296"
      r="2"
      transform="translate(0 11)"
    />
    <rect
      fill="#714601"
      height="10"
      id="Rectangle_1531"
      rx="1.5"
      transform="translate(0.5)"
      width="3"
    />
  </svg>
  <div
    class="mssg"
  >
    Message
  </div>
  <div
    class="closeBtn"
  >
    <svg
      height="10"
      id="Cancel"
      viewBox="0 0 15.021 15.021"
      width="7"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.374,14.5l-4.95-4.95L2.475,14.5A1.5,1.5,0,0,1,.354,12.374L5.3,7.425.354,2.475A1.5,1.5,0,0,1,2.475.354L7.425,5.3l4.95-4.95A1.5,1.5,0,0,1,14.5,2.475l-4.95,4.95,4.95,4.95A1.5,1.5,0,1,1,12.374,14.5Z"
        fill="#714601"
        id="Union_3"
        transform="translate(0.086 0.086)"
      />
    </svg>
  </div>
</div>
`;

exports[`<Alert /> matches snapshots 4`] = `
<div
  class="success success--light"
  data-testid="success"
  role="alert"
>
  <svg
    class="successIcon"
    height="13.397"
    viewBox="0 0 14.242 14.397"
    width="13.242"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      id="Component_74_1"
      transform="translate(1.299 0.444)"
    >
      <g
        id="Group_146"
        transform="translate(-498.31 -3301.875) rotate(-4)"
      >
        <line
          fill="none"
          id="Line_2"
          stroke="#215F0C"
          stroke-linecap="round"
          stroke-width="3"
          transform="translate(267.073 3336.275)"
          x2="3.427"
          y2="3.224"
        />
        <path
          d="M7.817-1.6,1.7,7.672"
          fill="none"
          id="Path_38"
          stroke="#215F0C"
          stroke-linecap="round"
          stroke-width="3"
          transform="translate(268.811 3331.831)"
        />
      </g>
    </g>
  </svg>
  <div
    class="mssg"
  >
    Message
  </div>
  <div
    class="closeBtn"
  >
    <svg
      height="10"
      id="Cancel"
      viewBox="0 0 15.021 15.021"
      width="7"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.374,14.5l-4.95-4.95L2.475,14.5A1.5,1.5,0,0,1,.354,12.374L5.3,7.425.354,2.475A1.5,1.5,0,0,1,2.475.354L7.425,5.3l4.95-4.95A1.5,1.5,0,0,1,14.5,2.475l-4.95,4.95,4.95,4.95A1.5,1.5,0,1,1,12.374,14.5Z"
        fill="#215F0C"
        id="Union_3"
        transform="translate(0.086 0.086)"
      />
    </svg>
  </div>
</div>
`;

exports[`<Alert /> matches snapshots 5`] = `
<div
  class="error error--light"
  data-testid="error"
  role="alert"
>
  <svg
    class="errorIcon"
    height="15.817"
    viewBox="0 0 11.3 15.817"
    width="11.3"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      id="error-fire"
      transform="translate(0 -0.1)"
    >
      <path
        d="M-542-3218.083a6.007,6.007,0,0,1-6-6c0-3.289,5.713-9.753,5.771-9.818a2.815,2.815,0,0,0,.23.486,19.62,19.62,0,0,0,1.149,1.788l.086.126.254.374c-.023.029-2.261,2.794-2.215,4.1a2.426,2.426,0,0,0,2.432,2.349h.131a2.227,2.227,0,0,0,1.6-.681,2.221,2.221,0,0,0,.616-1.654,8.461,8.461,0,0,1,1.249,3.6,5.4,5.4,0,0,1-1.407,3.91A5.329,5.329,0,0,1-542-3218.083Z"
        fill="#d01d23"
        id="Subtraction_16"
        transform="translate(548 3234)"
      />
    </g>
  </svg>
  <div
    class="mssg"
  >
    Message
  </div>
  <div
    class="closeBtn"
  >
    <svg
      height="10"
      id="Cancel"
      viewBox="0 0 15.021 15.021"
      width="7"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.374,14.5l-4.95-4.95L2.475,14.5A1.5,1.5,0,0,1,.354,12.374L5.3,7.425.354,2.475A1.5,1.5,0,0,1,2.475.354L7.425,5.3l4.95-4.95A1.5,1.5,0,0,1,14.5,2.475l-4.95,4.95,4.95,4.95A1.5,1.5,0,1,1,12.374,14.5Z"
        fill="#d01d23"
        id="Union_3"
        transform="translate(0.086 0.086)"
      />
    </svg>
  </div>
</div>
`;
